Separate legacy WC menu handling to a separate hook,

tomalec opened the pull request:
Tastes best with https://github.com/woocommerce/google-listings-and-ads/pull/1031 
### Changes proposed in this Pull Request:

Separate legacy WC menu handling to a separate hook, from `<MainTabNav>`.

Simplify it to use `window.wpNavMenuClassChange` instead of custom CSS selectors hacking.
Use it in `Settings` sub-pages.

Addresses part of https://github.com/woocommerce/google-listings-and-ads/issues/1037.


After this change, any page can highlight the menu, even ones without `<MainTabNav>`


### Screenshots:

#### Before 
![image](https://user-images.githubusercontent.com/17435/136079018-3b1e7fd3-4bb2-487c-bbb5-946d22fec41c.png)

#### After
![image](https://user-images.githubusercontent.com/17435/136077588-907833ac-7d25-4ffb-8e51-003a34a6cbf5.png)


### Detailed test instructions:

0. Make sure you have WC Navigation disabled
1. Go to Settings page [`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings%2Fedit-phone-number)
2. and its subpages:
	1. 	[`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings&subpath=%2Freconnect-accounts`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings&subpath=%2Freconnect-accounts)`
	2. [`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings%2Fedit-phone-number`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings%2Fedit-phone-number)/[`https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings&subpath=%2Fedit-phone-number`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings&subpath=%2Fedit-phone-number)

Make sure the "Google Listings & Ads" is highlighted on the left menu.


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Tweak - refactored legacy WC menu highlighting effect.
> Fix - Report tabs lose active state when changing chart.


### Additional notes
1. I had a problem with naming. We have the current WC navigation and the new one. I used "Legacy" for the current one, and "WC Navigation" for the new one, but I'm still not happy with those. "New", "old", "legacy", and any time-related name are almost by definition not stable in time. Every "new" would eventually become "old" and the current "new" may at some point become "legacy" leaving no way to differentiate the two. Any ideas for something better?

tomalec committed:
Separate legacy WC menu handling to a separate hook,

from `<MainTabNav>`.

Simplify it to use `window.wpNavMenuClassChange` instead of custom CSS selectors hacking.
Use it in `Settings` sub-pages.

Addresses part of https://github.com/woocommerce/google-listings-and-ads/issues/1037.

eason9487 reviewed (commented):
This method looks good. But I have two questions:

1. ❓  Although the active state of tabs already had a glitch (#621), the active state is completely disappeared in this PR.
2. ❔  I wonder if the issue that WC menu highlight disappears after repeatedly clicking on the same tab could be fixed by this method as well?
  ![Kapture 2021-10-08 at 19 28 18](https://user-images.githubusercontent.com/17420811/136549074-8dc9f684-5b1e-430d-958f-aecd4a83d0c2.gif)

tomalec commented:
> 2. I wonder if the issue that WC menu highlight disappears after repeatedly clicking on the same tab could be fixed by this method as well?

I got the same effect on `develop`. And to me, it seems to be related to the `<Link>` used in the sub-navigation https://github.com/woocommerce/google-listings-and-ads/issues/984.

WP-admin re-renders the menu on the navigation even, and we have no hook to react on that.

We could fix it my calling `useLegacyMenuEffect` in the reports navigation render. But Id' rather fix it in a wider scope of navigation-related issues https://github.com/woocommerce/google-listings-and-ads/issues/1037
or at least by changing `reportKey` to `path`, like we did for `subpath` and make the legacy menu handler simply react on any `{ path } = getQuery()` change.

tomalec commented:
> 1. Although the active state of tabs already had a glitch (#621), the active state is completely disappeared in this PR.

:man_facepalming:

This is a result of ignoring the scoping in WP functions 
```js
window.wpNavMenuClassChange = function ( page, url ) {
	Array.from( document.getElementsByClassName( 'current' ) ).forEach(
		function ( item ) {
			item.classList.remove( 'current' );
		}
	);

```
it removes `current` class from **every** element on the page.


<code>wp<strong>NavMenu</strong>ClassChange: … <strong>document</strong>.getElementsByClassName</code>
To me, it's a fundamental bug on WP/WC end. We should either leave it as is not to maintain too much hacky code on our end. Or introduce the actual CSS boundary on our end.

tomalec commented:
Created an issue for 1. at https://github.com/woocommerce/woocommerce-admin/issues/7772

tomalec commented:
I also fixed it in WC https://github.com/woocommerce/woocommerce-admin/pull/7773
Now, the question is should we keep this bug, or keep maintaining a fork as workaround until, our lowest supported WC version is the one that hopefully would have that fix. (my guesstimate is WC-admin 2.11.x => WC totally unknown 6.1?)

//cc @jconroy

eason9487 commented:
Looks like woocommerce/woocommerce-admin#7773 can fix both issues within the https://github.com/woocommerce/google-listings-and-ads/pull/1038#pullrequestreview-774899799. 💯 

> Now, the question is should we keep this bug, or keep maintaining a fork as workaround until, our lowest supported WC version is the one that hopefully would have that fix.

Finally found out where the `.subsubsub .current` style comes from.

- https://github.com/WordPress/wordpress-develop/blob/5.8.1/src/wp-admin/css/common.css#L448-L451
- https://github.com/WordPress/wordpress-develop/blob/5.8.1/src/wp-admin/css/common.css#L683-L685

Considering that 1) probably we still have a few months to wait for WC-admin to release the fixes and our L-2 version to be ready for bumping up, and 2) the cost looks small enough if we maintain a copy that has only 5 lines and has not been changed for 8 years. As a trade-off to exchange for fixing #621 early, I think it's okay.

tomalec committed:
Work around `current` class pruning,

https://github.com/woocommerce/woocommerce-admin/issues/7772.

Addresses https://github.com/woocommerce/google-listings-and-ads/issues/621, https://github.com/woocommerce/google-listings-and-ads/pull/1038#pullrequestreview-774899799

tomalec commented:
> Considering that 1) probably we still have a few months to wait for WC-admin to release the fixes and our L-2 version to be ready for bumping up, and 2) the cost looks small enough if we maintain a copy that has only 5 lines and has not been changed for 8 years. As a trade-off to exchange for fixing #621 early, I think it's okay.

Good point. I duplicated the `current`'s styles as `gla-sub-nav__item--current`'s ones.
![subnav](https://user-images.githubusercontent.com/17435/137120867-19400552-5ddd-45a4-b6b7-25e538445a7f.gif)

eason9487 reviewed (approved):
Testing well. LGTM.

tomalec commented:
I made an issue, to eventually remove the workaround: https://github.com/woocommerce/google-listings-and-ads/issues/1048.

and a follow up for the [2. above](https://github.com/woocommerce/google-listings-and-ads/pull/1038#pullrequestreview-774899799):  https://github.com/woocommerce/google-listings-and-ads/issues/1049
